src/transformers/models/big_bird/modeling_big_bird.py

```diff
@@ -0,0 +1,2695 @@
+
+        # Define shorthands
+        h = num_attention_heads
+        r = num_rand_blocks
+        rsqrt_d = 1 / math.sqrt(attention_head_size)
+        b = batch_size
+        m = from_seq_length
+        n = to_seq_length
+        wm = from_block_size
+        wn = to_block_size
+
+        # generate random attention and corresponding masks
+        np.random.seed(seed)
+        if from_seq_length in [1024, 3072, 4096]:  # old plans used in paper
+            rand_attn = [
+                self._bigbird_block_rand_mask(self.max_seqlen, self.max_seqlen, wm, wn, r, last_idx=1024)[
+                    : (from_seq_length // wm - 2)
+                ]
+                for _ in range(h)
+            ]
+        else:
+            if plan_from_length is None:
+                plan_from_length, plan_num_rand_blocks = self._get_rand_attn_plan(from_seq_length, wm, r)
+
+            rand_attn = self._bigbird_block_rand_mask_with_head(
+                from_seq_length=from_seq_length,
+                to_seq_length=to_seq_length,
+                from_block_size=wm,
+                to_block_size=wn,
+                num_heads=h,
+                plan_from_length=plan_from_length,
+                plan_num_rand_blocks=plan_num_rand_blocks,
+            )
+
+        rand_attn = np.stack(rand_attn, axis=0)
+        rand_attn = torch.tensor(rand_attn, device=query_layer.device, dtype=torch.long)
+        rand_attn.unsqueeze_(0)
+        rand_attn = torch.cat([rand_attn for _ in range(batch_size)], dim=0)
+
+        rand_mask = self._create_rand_mask_from_inputs(from_blocked_mask, to_blocked_mask, rand_attn, h, r, b, m, wm)
+
+        blocked_query_matrix = query_layer.view(b, h, m // wm, wm, -1)
+        blocked_key_matrix = key_layer.view(b, h, n // wn, wn, -1)
+        blocked_value_matrix = value_layer.view(b, h, n // wn, wn, -1)
+
+        # preparing block for randn attn
+        gathered_key = self.torch_gather_b2(blocked_key_matrix, rand_attn)
+        gathered_key = gathered_key.view(b, h, n // wn - 2, r * wn, -1)  # [b, h, n//wn-2, r, wn, -1]
+        gathered_value = self.torch_gather_b2(blocked_value_matrix, rand_attn)
+        gathered_value = gathered_value.view(b, h, n // wn - 2, r * wn, -1)  # [b, h, n//wn-2, r, wn, -1]
+
+        # 1st block is global q[0] x (k[0], k[1], k[2], k[3], k[4] .... )
+
+        # [b, h, wm, -1] x [b, h, n, -1] ==> [b, h, wm, n]
+        first_product = self.torch_bmm_nd_transpose(blocked_query_matrix[:, :, 0], key_layer, ndim=4)
+
+        first_product = first_product * rsqrt_d
+        first_product += (1.0 - to_mask) * -10000.0
+        first_attn_weights = F.softmax(first_product, dim=-1)  # [b, h, wm, n]
+
+        # [b, h, wm, n] x [b, h, n, -1] ==> [b, h, wm, -1]
+        first_context_layer = self.torch_bmm_nd(first_attn_weights, value_layer, ndim=4)
+        first_context_layer.unsqueeze_(2)
+
+        # q[1] x (sliding_keys, random_keys, global_keys)
+
```

Comment:
maybe a short comment here as well saying what this is about (the second block?)